### ISSUE TITLE
Add grunt deploy-phet-io command

### ISSUE THREAD

**Issue by samreid** [phetsims/chipper#481]
Currently after a sim is built the developer must do the following steps to deploy to phet-server:
https://github.com/phetsims/phet-info/blob/master/deployment-info/sim_deployment.md

ssh into phet-server.int (requires VPN)
scp -r spot:/htdocs/physics/phet/dev/html/SIMULATION/VERSION /data/web/static/phet-io/0.1/sims/SIMULATION/VERSION
cp /data/web/static/phet-io/0.1/sims/concentration/1.3.0-dev.3/phet-io/protected/.htaccess /data/web/static/phet-io/0.1/sims/SIMULATION/VERSION/phet-io/protected/.htaccess

We would rather have a `grunt deploy-phet-io` command that does all of this.  It could be implemented like `grunt deploy-dev`.

**Comment by samreid** [comment]
I can see at least 2 "right" ways to do this:

1. Generalize deployDev so it can deploy to phet-server
2. Use build-server to deploy to phet-server.

I'm not familiar enough with the build process to make the call here.  Can @jbphet @pixelzoom @andrew @jonathanolson @jessegreenberg or @mattpen chime in?  Tagging for developer-meeting in case we don't get it sorted out by then.

**Comment by jonathanolson** [comment]
If it's a production release, I'd lean towards build-server, so the builds are consistent (same single platform/setup used to build all production sims).

**Comment by samreid** [comment]
Yes, phet-io deploys are for production (though until now we have used dev versioning), so I'm up for using build-server.  I haven't worked with it before, perhaps it would be useful for someone with build-server experience to help out on this part.

**Comment by jbphet** [comment]
+1 for using the build server.  For non-phet-io deployments, the build server is used for release candidate and production builds, but not for dev deployments.  It would probably be best to be consistent.

**Comment by pixelzoom** [comment]
@samreid wrote:
> I can see at least 2 "right" ways to do this:
> 
> (1) Generalize deployDev so it can deploy to phet-server

While it may currently be the case that PhET-iO deployments are dev versions, that's undesirable in the long run.  Presumably PhET-iO will have both production and dev versions.

> (2) Use build-server to deploy to phet-server.

That depends on whether you're deploying for production, RC, or dev. For production and RC, use build-server.  For dev, build locally. 

Ultimately, we should be able to use `deploy-production`, `deploy-RC` and `deploy-dev` grunt tasks to build any brand of PhET sim, including phet-io.

**Comment by pixelzoom** [comment]
@samreid wrote:
> Yes, phet-io deploys are for production (though until now we have used dev versioning),

Imo, we should stop using dev versioning for phet-io deploys asap.

**Comment by samreid** [comment]
@ariel-phet suggested @samreid could work with @mattpen on using the build server.

**Comment by samreid** [comment]
@mattpen and I discussed this, and we have the following recommended steps:

Steps for chipper side:
* Change the instructions for building a phet-io branded simulation so that it uses `grunt --brand=phet-io` instead of `grunt build-phet-io`
* Add a step to the phet-io build & deployment guidelines that builds the phet branded one as well (if we are still doing that)
* Add a step to afterRequireJSBuild.js that checks if the brand is phet-io, then it runs the logic which is currently in buildPhETIO.js
* Pass the brand through the query parameters to the build server

Steps for build-server side:
* make the brand an argument which is passed over query parameters and respected in the `grunt build --brand=...` step
* Make sure the SCP Spot command will work OK for phet-io RC deploys, including the .htaccess password protection.
* Add an if statement for brand==='phet' and brand==='phet-io' that puts the files in the right location.

**Comment by samreid** [comment]
To Do:

- [ ] build-phet-io directory should not be created
- [ ] pop up lib/ and protected/ one level

**Comment by samreid** [comment]
> Pass the brand through the query parameters to the build server

If the phet-io versions have the string "phetio." in them, that can be checked on the server.

**Comment by mattpen** [comment]
Todo items for the build server:

- [ ] Write an .htaccess file in the spotScp for phet-io sims
- [ ] Make sure that the scp from phet-server to spot is recursive for phet-io sims.
- [ ] Check that the .htaccess file is copied with the scp.

**Comment by samreid** [comment]
Chipper-side TODOs are complete.

**Comment by samreid** [comment]
Our new task is lacking the ability to do dev-deploys.  Can we use RC process with a dev version?  Or if we want to use deployDev with phet-io, we will have to augment with .htaccess.

Why aren't we using the build server for dev builds?

**Comment by mattpen** [comment]
@samreid - The deploy-rc task successfully built the sim and deployed the files to spot.  I noticed during a brief check that the `screenshot` wrapper is working, but that [`instance-proxies`](http://www.colorado.edu/physics/phet/dev/html/faradays-law/1.2.0-phetio.1/protected/wrappers/instance-proxies/instance-proxies.html) is saying that all values in the sim are undefined.

**Comment by samreid** [comment]
What is the ox-dev URL for the deployed sim?

**Comment by samreid** [comment]
The error looks like this:

![image](https://cloud.githubusercontent.com/assets/679486/18750245/9318e74e-8097-11e6-84c5-8365fc015016.png)

**Comment by samreid** [comment]
The problem goes away with `?launchLocalVersion`

**Comment by samreid** [comment]
I wonder if we should always `launchLocalVersion` with a fallback to loading the one from phet-io.colorado.edu for clients?

**Comment by samreid** [comment]
In the preceding commit, I removed the 0.1 from the path in wrappers.  @mattpen you will need to pull before doing a production phet-io deploy.

**Comment by samreid** [comment]
@mattpen and I made a lot of progress on this, I am not sure where it is at.  @mattpen what remains to be done here?

**Comment by mattpen** [comment]
I tested a production deploy of both phet and phet-io branded sims to ox-dev tonight.  The phet branded sim can be seen here: https://ox-dev.colorado.edu/sims/html/faradays-law/latest/faradays-law_en.html (albeit with a phet-io version string).  The phet-io brand can be checked via ssh at `phet-server-dev:/data/web/static/phet-io/sims/faradays-law`.  The apache directives to display phet-io-dev.colorado.edu have not been setup yet.  

@samreid - Should we progress with a test on phet-server, or would it be better to get the phet-io-dev site working to test the phet-io sim and wrappers in a a browser first?

**Comment by samreid** [comment]
> The apache directives to display phet-io-dev.colorado.edu have not been setup yet.

That seems like something we want to do anyways.  When will it be ready and does it require OIT intervention?

I like the process of testing things on phet-dev first, but if it will be too long before it is ready we can go straight to production.  There are no "update notification" or disturbances for our clients if we deploy extra phet-io things onto phet-server, so that seems like it would be OK.

**Comment by mattpen** [comment]
> When will it be ready and does it require OIT intervention?

I can get it ready in 3-4 hours, I can work on it this weekend if you'd like to have it working before testing in production.

It does not require any intervention by OIT (unless they incorrectly assigned the IP address for phet-io-dev), but it might require some help from @phetsims/mark-morlino.

**Comment by mattpen** [comment]
@samreid - I'm still working out a few details, but you can see the deployed files in a browser at https://128.138.129.31/sims/faradays-law/.  This requires a VPN connection.

**Comment by samreid** [comment]
Looks great, I don't see any trouble there.  The wrappers need ?launchLocalVersion unless using the version on phet-io.colorado.edu, in order to test them.  Perhaps we should come up with a better way of doing that, but all else seems good.

**Comment by samreid** [comment]
@mattpen So is the next step to test a production deploy?

**Comment by mattpen** [comment]
The dev domain name is working now https://phet-io-dev.colorado.edu/sims/, although we are missing a cert, so you'll have to tell your browser to ignore the SSL error.

I think a good next step would be to deploy a rosetta translation to the tests babel branch using ox-dev.colorado.edu, to make sure that rosetta still works after our changes.  I'll work on this tonight.

**Comment by mattpen** [comment]
From phetsims/phet-io#638:
>mattpen said: samreid - zepumph and I deployed the latest phet-io version of BLL to ox-dev today, it is available https://phet-io-dev.colorado.edu/sims/beers-law-lab. However, the lab-book wrapper is not able to launch because it is looking for the sim in this directory: /sims/beers-law-lab/1.5.0-phetio.4/phet-io/.
>
>Should we remove the /phet-io/ directory from PhETIOAppUtils?

>samreid said: Yes, I can remove that path and will do so shortly. However, we need ?launchLocalVerison to test away from phet-io. The wrappers are coded to launch the version on phet-io.colorado.edu at the moment. If there's nothing there, they will error out, unless ?launchLocalVersion tells them to look at relative paths.


> samreid said: OK I removed the phet-io subdir, but please see my preceding comment about why this won't fix things on ox-dev.

**Comment by mattpen** [comment]
@samreid - somehow that path is being interpreted as relative on phet-io-dev.  I did another deploy after the fix and it is launching correctly without the ?launchLocalVersion parameter.  See https://phet-io-dev.colorado.edu/sims/beers-law-lab/1.5.0-phetio.4/protected/wrappers/lab-book/lab-book.html?sim=beers-law-lab.  

I would suspect that this may be because Apache thinks that phet-io.colorado.edu belongs to localhost?

**Comment by mattpen** [comment]
@samreid - I deployed a phetio version of FB to phet-server tonight - https://phet-io.colorado.edu/sims/function-builder/1.1.0-phetio.4/function-builder_en-phetio.html.  It is unable to launch wrappers because it is dependent on an old version of chipper that still included `/phet-io/` in the path.  Should I fix this manually, or do you foresee this being a common occurrence that we should handle during deployment?

**Comment by mattpen** [comment]
I added a rewrite rule that redirects all requests to /sims/SIM/VERSION/phet-io/* to /sims/SIM/VERSION/*.  It appears to be working for function builder, see https://phet-io.colorado.edu/sims/function-builder/1.1.0-phetio.4/protected/wrappers/instance-proxies/instance-proxies.html

**Comment by samreid** [comment]
>  It is unable to launch wrappers because it is dependent on an old version of chipper that still included `/phet-io/` in the path.

Can you help me understand the problem?  In what way was function builder dependent on an old version of chipper?  Was using release-candidate SHAs or something?

> I added a rewrite rule that redirects all requests to /sims/SIM/VERSION/phet-io/* to /sims/SIM/VERSION/*.

Wouldn't that break links we have already given to customers for historical versions that do use the phet-io/ paths?

**Comment by mattpen** [comment]
> Wouldn't that break links we have already given to customers for historical versions that do use the phet-io/ paths?  

No, as these historical links are all in the /0.1/ directory, which the rewrite rule would not match.  This problem is addressed further in phetsims/phet-io-website#80.

>Can you help me understand the problem? In what way was function builder dependent on an old version of chipper? Was using release-candidate SHAs or something?

Function builder is using the following SHA for chipper:
```
"chipper": {
    "sha": "2692cf6736e2bc304659c961e1cf357b870e49b0",
    "branch": "HEAD"
  }
```
This sha is from before the phet-io subdirectory was modified for PhETIOAppUtils.

**Comment by samreid** [comment]
> Function builder is using the following SHA for chipper:

We should start using new SHAs for new phet-io releases, so that the chipper build process (among other things) will be correct.

I'm not sure what other feedback to give on this issue, if any.  Are there any open questions?  Anything else to do?

**Comment by mattpen** [comment]
The issue with conflicting SHAs was resolved here: phetsims/phet-io-website#80
